// @gilsdav/ngx-translate-router-http-loader/esm2015/lib/http-loader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { LocalizeParser } from '@gilsdav/ngx-translate-router';
/**
 * Config interface
 * @record
 */
export function ILocalizeRouterParserConfig() { }
if (false) {
    /** @type {?} */
    ILocalizeRouterParserConfig.prototype.locales;
    /** @type {?|undefined} */
    ILocalizeRouterParserConfig.prototype.prefix;
}
export class LocalizeRouterHttpLoader extends LocalizeParser {
    /**
     * CTOR
     * @param {?} translate
     * @param {?} location
     * @param {?} settings
     * @param {?} http
     * @param {?=} path
     */
    constructor(translate, location, settings, http, path = 'assets/locales.json') {
        super(translate, location, settings);
        this.http = http;
        this.path = path;
    }
    /**
     * Initialize or append routes
     * @param {?} routes
     * @return {?}
     */
    load(routes) {
        return new Promise((resolve) => {
            this.http.get(`${this.path}`)
                .subscribe((data) => {
                this.locales = data.locales;
                this.prefix = data.prefix || '';
                this.init(routes).then(resolve);
            });
        });
    }
}
if (false) {
    /** @type {?} */
    LocalizeRouterHttpLoader.prototype.http;
    /** @type {?} */
    LocalizeRouterHttpLoader.prototype.path;
}
