// @gilsdav/ngx-translate-router/esm5/lib/gilsdav-reuse-strategy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var GilsdavReuseStrategy = /** @class */ (function () {
    // private handlers: {[key: string]: DetachedRouteHandle} = {};
    function GilsdavReuseStrategy() {
    }
    /**
     * @param {?} route
     * @return {?}
     */
    GilsdavReuseStrategy.prototype.shouldDetach = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        // console.log('shouldDetach', route);
        return false;
    };
    /**
     * @param {?} route
     * @param {?} handle
     * @return {?}
     */
    GilsdavReuseStrategy.prototype.store = /**
     * @param {?} route
     * @param {?} handle
     * @return {?}
     */
    function (route, handle) {
        // console.log('store', route, handle);
        // console.log('store url', this.getKey(route));
        // this.handlers[this.getKey(route)] = handle;
    };
    /**
     * @param {?} route
     * @return {?}
     */
    GilsdavReuseStrategy.prototype.shouldAttach = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        // console.log('shouldAttach', route, this.getKey(route));
        // return !!this.handlers[this.getKey(route)];
        return false;
    };
    /**
     * @param {?} route
     * @return {?}
     */
    GilsdavReuseStrategy.prototype.retrieve = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        // console.log('retrieve', route);
        // console.log('retrieve url', this.getKey(route));
        // const result = this.handlers[this.getKey(route)];
        // delete this.handlers[this.getKey(route)];
        // return result;
        return null;
    };
    /**
     * @param {?} future
     * @param {?} curr
     * @return {?}
     */
    GilsdavReuseStrategy.prototype.shouldReuseRoute = /**
     * @param {?} future
     * @param {?} curr
     * @return {?}
     */
    function (future, curr) {
        // console.log('shouldReuseRoute', future, curr, this.getKey(future) === this.getKey(curr));
        // console.log('shouldReuseRoute', future && curr ? this.getKey(future) === this.getKey(curr) : false);
        return future && curr ? this.getKey(future) === this.getKey(curr) : false;
    };
    /**
     * @param {?} route
     * @return {?}
     */
    GilsdavReuseStrategy.prototype.getKey = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        // console.log(route.parent.component.toString());
        if (!route.data.localizeRouter && (!route.parent || !route.parent.parent)) {
            return 'LANG';
        }
        else if (route.data.localizeRouter) {
            return this.getKey(route.parent) + "/" + route.data.localizeRouter.path;
        }
    };
    return GilsdavReuseStrategy;
}());
export { GilsdavReuseStrategy };
